tests/notebooks/test_run_notebooks.py

```diff
@@ -1,54 +1,295 @@
 
-# TODO: These will be replaced with stronger tests using testbook soon
-def _test_run_notebook(folder, notebook_fname, n_cells):
-    # change to notebook directory to allow testing
-    cwd = os.getcwd()
+
+# return testbook for given notebook
+def open_book(folder, notebook_fname, **kwargs):
+    execute = kwargs.get("execute", True)
     os.chdir(os.path.join(this_file_dir(), "..", "..", "docs", "notebooks", folder))
-    with testbook(notebook_fname, timeout=300, execute=True) as tb:
-        assert tb.code_cells_executed == n_cells
-    os.chdir(cwd)
+    book = testbook(notebook_fname, execute=execute, timeout=300)
+    return book
+
+
+# checks that the number of executed cells matches the expected
+def check_cell_execution(tb, notebook_fname, **kwargs):
+    injections = kwargs.get("injections", 0)
+    assert (
+        tb.code_cells_executed
+        == cell_counter(notebook_fname, only_code_cells=True) + injections
+    )
+
+
+# checks for correct type and number of layers in a model
+def check_layers(tb, activations, network):
+    tb.inject(
+        f"""
+                    activations = {activations}
+                    for layer_id, layer in enumerate({network}):
+                        assert activations[layer_id] in str(layer.activation)
+                """
+    )
+
+
+# counting number of cells
+def cell_counter(notebook_fname, **kwargs):
+    only_code_cells = kwargs.get("only_code_cells", False)
+    nb = nbformat.read(notebook_fname, as_version=4)
+    nb = nbformat.validator.normalize(nb)[1]
+    if only_code_cells:
+        total = 0
+        for cell in nb.cells:
+            if cell["cell_type"] == "code" and len(cell["source"]) != 0:
+                total += 1
+        return total
+    else:
+        return len(nb.cells)
+
+
+# gets model stats for mnist notebooks
+def mnist_stats(tb, fname):
+    total_cells = cell_counter(fname)
+    tb.inject("test(model, test_loader)")
+    model_stats = tb.cell_output_text(total_cells)
+    model_stats = model_stats.split(" ")
+    loss = float(model_stats[4][:-1])
+    accuracy = int(model_stats[-2][:-6])
+    return (loss, accuracy)
+
+
+# neural network formulation notebook helper
+def neural_network_checker(tb, ref_string, val1, val2, tolerance):
+    x = tb.ref(f"{ref_string}[0]")
+    y = tb.ref(f"{ref_string}[1]")
+    assert x == pytest.approx(val1, abs=tolerance)
+    assert y == pytest.approx(val2, abs=tolerance)
 
 
 @pytest.mark.skipif(not keras_available, reason="keras needed for this notebook")
 def test_autothermal_relu_notebook():
-    _test_run_notebook("neuralnet", "auto-thermal-reformer-relu.ipynb", 13)
+    notebook_fname = "auto-thermal-reformer-relu.ipynb"
+    book = open_book("neuralnet", notebook_fname)
+
+    with book as tb:
+        check_cell_execution(tb, notebook_fname)
+
+        # check loss of model
+        model_loss = tb.ref("nn.evaluate(x, y)")
+        assert model_loss == pytest.approx(0.000389626, abs=0.00031)

```

Comment:
The model loss will not be consistent enough to check this way. I think we should just check if it is at least as small as some acceptable tolerance - just so we know if the network begins training more poorly for some reason.